simx86: disassemble JIT-generated code for 'e' level>7

bartoldeman opened the pull request:
Some very basic changes were made to dis8086.c so it understands JIT-generated x86-64 code.

Output looks like this

```
============ Opening sequence at 000ffff0
  000ffff0: ea5be000f0        ffff:0000 jmp  F000:E05B
CGEN:  13  40001
adding ff000 to cache
CGEN:  13  40003
CGEN:  13  40001
CGEN:  13  40001
CGEN: 114 140003
JMP_FAR: f000:0000e05b
==== Closing sequence at 000ffff5
---------------------------------------------
ProduceCode: CurrIMeta=0
CodeBuf=0x7f6868e8e3c8 siz 380
  0x7f6868e8e3c8: c7432c00000000        mov  dword [rbx+002C],0
PGEN(00,00)  13  40001  7 0000002c 00000000 00000000 00000000 00000000
  0x7f6868e8e3cf: 66c7433600f0          mov  word [rbx+0036],F000
PGEN(00,01)  13  40003  6 00000036 0000f000 00000000 00000000 00000000
  0x7f6868e8e3d5: c7434800000f00        mov  dword [rbx+0048],f0000
PGEN(00,02)  13  40001  7 00000048 000f0000 00000000 00000000 00000000
  0x7f6868e8e3dc: c7434cffff0f00        mov  dword [rbx+004C],fffff
PGEN(00,03)  13  40001  7 0000004c 000fffff 00000000 00000000 00000000
JMP_Link 000fe05b 000ffff0
  0x7f6868e8e3e3: 0fb74b72              movzx ecx,[rbx+0072]
  0x7f6868e8e3e7: e309                  jecxz 7f6868e8e3f2 ($+9)
  0x7f6868e8e3e9: b85be00f00            mov  eax,fe05b
  0x7f6868e8e3ee: 89c1                  mov  ecx,eax
  0x7f6868e8e3f0: 5a                    pop  edx
  0x7f6868e8e3f1: c3                    ret
  0x7f6868e8e3f2: b9f0ff0f00            mov  ecx,ffff0
  0x7f6868e8e3f7: b85be00f00            mov  eax,fe05b
  0x7f6868e8e3fc: 5a                    pop  edx
  0x7f6868e8e3fd: c3                    ret
PGEN(00,04) 114 140003 27 000fe05b 000ffff0 0000001b 00000000 00000000
```

bartoldeman committed:
simx86: disassemble JIT-generated code for 'e' level>7

Some very basic changes were made to dis8086.c so it understands
JIT-generated x86-64 code.

Output looks like this

```
============ Opening sequence at 000ffff0
  000ffff0: ea5be000f0        ffff:0000 jmp  F000:E05B
CGEN:  13  40001
adding ff000 to cache
CGEN:  13  40003
CGEN:  13  40001
CGEN:  13  40001
CGEN: 114 140003
JMP_FAR: f000:0000e05b
==== Closing sequence at 000ffff5
---------------------------------------------
ProduceCode: CurrIMeta=0
CodeBuf=0x7f6868e8e3c8 siz 380
  0x7f6868e8e3c8: c7432c00000000        mov  dword [rbx+002C],0
PGEN(00,00)  13  40001  7 0000002c 00000000 00000000 00000000 00000000
  0x7f6868e8e3cf: 66c7433600f0          mov  word [rbx+0036],F000
PGEN(00,01)  13  40003  6 00000036 0000f000 00000000 00000000 00000000
  0x7f6868e8e3d5: c7434800000f00        mov  dword [rbx+0048],f0000
PGEN(00,02)  13  40001  7 00000048 000f0000 00000000 00000000 00000000
  0x7f6868e8e3dc: c7434cffff0f00        mov  dword [rbx+004C],fffff
PGEN(00,03)  13  40001  7 0000004c 000fffff 00000000 00000000 00000000
JMP_Link 000fe05b 000ffff0
  0x7f6868e8e3e3: 0fb74b72              movzx ecx,[rbx+0072]
  0x7f6868e8e3e7: e309                  jecxz 7f6868e8e3f2 ($+9)
  0x7f6868e8e3e9: b85be00f00            mov  eax,fe05b
  0x7f6868e8e3ee: 89c1                  mov  ecx,eax
  0x7f6868e8e3f0: 5a                    pop  edx
  0x7f6868e8e3f1: c3                    ret
  0x7f6868e8e3f2: b9f0ff0f00            mov  ecx,ffff0
  0x7f6868e8e3f7: b85be00f00            mov  eax,fe05b
  0x7f6868e8e3fc: 5a                    pop  edx
  0x7f6868e8e3fd: c3                    ret
PGEN(00,04) 114 140003 27 000fe05b 000ffff0 0000001b 00000000 00000000
```

bartoldeman commented:
actually I'll make the disasm a little better, as replacing e by r will also mangle hex values.

bartoldeman commented:
done..

stsp commented on src/base/emu-i386/simx86/codegen-x86.c:
```diff
@@ -2211,12 +2212,11 @@ arith0:		{
 		    {
 			if ((ra > -127) && (ra < 128)) {
-			    ra -= 1; G1(JMPsid,Cp);
+			    ra -= 1; G1(JMPsid,Cp); G1(ra,Cp);
```
Why changes to codegen?
They are not changelogged,
what does this do?

bartoldeman replied:
I should split this out indeed.
It merely keeps Cp at the right place so the disassembler is not confused and tries to disas after the short jump but since it's a patch from the nodelinker does not affect allocation.

I'll split in a few hours.

bartoldeman replied:
ok, it's split now

bartoldeman committed:
simx86: only generate 1 byte for short jumps

This is the right thing to do, but before it did not
matter since the return Cp is ignored by the node linker
patching in the jmp.

Now the JIT disasm gets confused by it and would try to
disassemble the 3 bytes after the short jmp which are
nonsense.

stsp commented:
Thanks!
